test/blackbox-tests/test-cases/variables-for-artifacts/run.t

```diff
@@ -0,0 +1,190 @@
+  File "dune", line 39, characters 0-38:
+  39 | (alias
+  40 |  (name t5)
+  41 |  (deps %{cma:bar_}))
+  Error: No rule found for bar_.cma
+  [1]
+
+Command line version.
+
+  $ dune build %{cma:bar_}
+  Error: Don't know how to build bar_.cma
+  [1]
+
+This test tries to build a .cma in a subdirectory, where a different project is
+defined. The library is public in this case, but we use the local name.
+
+  $ dune clean
+  $ dune build --display short @t6
+        ocamlc sub2/.bar2.objs/byte/bar2.{cmi,cmo,cmt}
+      ocamldep sub2/.bar2.objs/y2.ml.d
+        ocamlc sub2/.bar2.objs/byte/bar2__Y2.{cmi,cmo,cmt}
+        ocamlc sub2/bar2.cma
+
+Command line version.
+
+  $ dune build --verbose %{cma:sub2/bar2} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/sub2/bar2.cma
+
+This test builds a .cmo in a subdirectory (same project).
+
+  $ dune clean
+  $ dune build --display short @t7
+      ocamldep sub/.bar.objs/x.ml.d
+        ocamlc sub/.bar.objs/byte/bar.{cmi,cmo,cmt}
+        ocamlc sub/.bar.objs/byte/bar__X.{cmi,cmo,cmt}
+
+Command line version.
+
+  $ dune build --verbose %{cmo:sub/x} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/sub/.bar.objs/byte/bar__X.cmo
+
+This test builds a module in a subdirectory (differnet project) belonging to a
+private library.
+
+  $ dune clean
+  $ dune build --display short @t8
+      ocamldep sub3/.c1.objs/x.ml.d
+        ocamlc sub3/.c1.objs/byte/c1.{cmi,cmo,cmt}
+        ocamlc sub3/.c1.objs/byte/c1__X.{cmi,cmo,cmt}
+
+COmmand line version.
+
+  $ dune build --verbose %{cmo:sub3/x} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/sub3/.c1.objs/byte/c1__X.cmo
+
+This test builds a private library in a subdirectory belonging to a different
+project.
+
+  $ dune clean
+  $ dune build --display short @t9
+        ocamlc sub3/.c1.objs/byte/c1.{cmi,cmo,cmt}
+      ocamldep sub3/.c1.objs/x.ml.d
+        ocamlc sub3/.c1.objs/byte/c1__X.{cmi,cmo,cmt}
+        ocamlc sub3/c1.cma
+
+Command line version.
+
+  $ dune build --verbose %{cma:sub3/c1} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/sub3/c1.cma
+
+This test builds a library in the current directory that has the same name as a
+public library defined in a subdirectory.
+
+  $ dune clean
+  $ dune build --display short @t10
+      ocamldep .c1.objs/c.ml.d
+        ocamlc .c1.objs/byte/c.{cmi,cmo,cmt}
+        ocamlc c1.cma
+
+Command line version.
+
+  $ dune build --verbose %{cma:c1} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/c1.cma
+
+This test checks error handling.
+
+  $ dune build %{cma:../x}
+  Error: path outside the workspace: ../x.cma from .
+  [1]
+
+This test checks that everything still works if we invoke dune from a
+subdirectory.
+
+  $ cd sub && dune build --display short %{cmx:x}
+  Info: Creating file dune-project with this contents:
+  | (lang dune 1.11)
```

Comment:
I don't understand why invoking `dune` from the `sub` subdirectory makes it create a `dune-project` file there; there is already one in the parent directory.

Reply:
When running dune inside dune, we disable root detection. It's like passing `--root .`. It's to avoid the inner dune from crawling up the filesystem outside of the test.

Reply:
Got it, thanks.